README.md
# Entrata for Gravity Forms

This plugin adds feeds capabilities for Entrata, using the [sendLeads API for Entrata](https://www.entrata.com/api/v1/documentation/sendLeads). The methodology used is standard, allowing for you to link up the basic custom fields in Entrata with your form fields using a Gravity Forms feed.

## Supported fields

The following fields are supported in this implementation:
* First name
* Last name
* Email
* Phone
* Message

## Before you begin

You will need to supply the following information on the backend of the site when setting up your feed:
* Entrata endpoint URL (the place where we send requests)
* Username
* Password
* Property ID
* Lead source ID

## Issues

If you're having trouble using this, try using a free service like [PipeDream](https://pipedream.com/) to test your requests to get better visibility into what you're seeing. Failing that, you can enable logging for the Entrata Gforms addon, then check what responses you're getting from the Entrata API.
 


CHANGES.md
### 1.0.2
* Adding logic to get the time in MT and submit it that way

### 1.0.1
* Adding some links for future reference

### 1.0
* Fully functional requests being made with all information going in from the backend
* Logging capabilities added
* Plugin updates capability added

### 0.1
* Request working successfully for sample form
* No functionality set up yet for feeds

lib/form-submit.php
<?php

// Set the propery ID
$property_id = '1058116';

//* Send the request headers for form ID 1
add_filter( 'gform_webhooks_request_headers_1', 'set_headers', 10, 4 );
function set_headers( $request_headers, $feed, $entry, $form ){
	
	//* This is the username:password but base64 encoded
    $request_headers['Authorization'] = 'Basic YnJpbmRsZV9hcGlAY2FyZGluYWw6QnJpbmRsZTUyMjIwMjBe';
    return $request_headers;
}

//* Set up the body for form ID 1
add_filter( 'gform_webhooks_request_data_1', 'modify_data_test_webhook_form', 10, 4 );
function modify_data_test_webhook_form( $request_data, $feed, $entry, $form ){    
	
	$data = $request_data;
	
	// Property details
	$property_id = '1058114';
	$firstname = $data['1.3'];
	$lastname = $data['1.6'];
	$email = $data['2'];
	$phone = $data['3'];
	$message = $data['4'];
	$date = date( 'm/d/Y') . 'T' . date( 'H:i:s' );
	$date = date( 'm/d/Y') . 'T' . '00:00:00';
	$leadsource = '41474'; // 41474 is the Community Website ID
	
	$request_data = [
		"auth" => [
			"type" => "basic" 
		], 
		"requestId" => "15", 
		"method" => [
			"name" => "sendLeads", 
			"params" => [
				"propertyId" => $property_id, 
				"doNotSendConfirmationEmail" => "1", 
				"isWaitList" => "0", 
				"prospects" => [
					"prospect" => [
						[
							"leadSource" => [
								"originatingLeadSourceId" => $leadsource, 
								// "additionalLeadSourceIds" => "xxxx,xxxx,xxxx" 
							], 
							"createdDate" => $date,
							// "createdDate" => "mm/dd/yyyyT03:47:33", 
							// "leasingAgentId" => "318487", 
							"customers" => [
								"customer" => [
									[
										"name" => [
											"firstName" => $firstname,
											"lastName" => $lastname, 
											// "namePrefix" => "Name Prefix", 
											// "middleName" => "Middle Name", 
											// "maidenName" => "Maiden Name", 
											// "nameSuffix" => "Name Suffix" 
										], 
										// "customerRelationshipTypeId" => "XXXX", 
										// "address" => [
										// 	"addressLine1" => "addressLine1", 
										// 	"addressLine2" => "addressLine2", 
										// 	"addressLine3" => "addressLine3", 
										// 	"city" => "city", 
										// 	"state" => "state code", 
										// 	"postalCode" => "postal code" 
										// ], 
										"phone" => [
											"personalPhoneNumber" => $phone, 
											// "cellPhoneNumber" => "xxx-xxx-xxxx", 
											// "officePhoneNumber" => "xxx-xxx-xxxx", 
											// "faxNumber" => "xxx-xxx-xxxx" 
										], 
										"email" => $email,
									] 
								] 
							], 
							"customerPreferences" => [
								// "desiredMoveInDate" => "mm/dd/yyyy", 
								// "desiredFloorplanId" => "xxxx", 
								// "desiredUnitTypeId" => "xxxx", 
								// "desiredUnitId" => "xxxx", 
								// "desiredRent" => [
								// 	"min" => "200", 
								// 	"max" => "300" 
								// ], 
								// "desiredNumBedrooms" => "2", 
								// "desiredNumBathrooms" => "3", 
								// "desiredLeaseTerms" => "18", 
								// "numberOfOccupants" => "1", 
								"comment" => $message, 
							], 
						// 	"events" => [
						// 		"event" => [
						// 			[
						// 				"date" => "mm/dd/yyyyT03:47:33", 
						// 				"callData" => [
						// 					"callFrom" => "xxx-xxx-xxxx", 
						// 					"ringThrough" => "xxx-xxx-xxxx", 
						// 					"callStatus" => "Answer", 
						// 					"duration" => "900", 
						// 					"audioLink" => "URL for the call" 
						// 				] 
						// 			], 
						// 			[
						// 				"type" => "Appointment", 
						// 				"date" => "mm/dd/yyyyT22:11:34", 
						// 				"appointmentDate" => "mm/dd/yyyy", 
						// 				"timeFrom" => "11:00am", 
						// 				"timeTo" => "11:30am", 
						// 				"eventReasons" => "Appointment Title", 
						// 				"comments" => "Appointment Description" 
						// 			], 
						// 			[
						// 				"type" => "EmailFromProspect", 
						// 				"date" => "mm/dd/yyyyT22:11:34", 
						// 				"emailAddresses" => [
						// 					"from" => "test1@email.com", 
						// 					"to" => "test11@email.com", 
						// 					"cc" => "test11@email.com", 
						// 					"bcc" => "test11@email.com" 
						// 				], 
						// 				"subject" => "Email Subject", 
						// 				"emailBody" => "Email Body" 
						// 			], 
						// 			[
						// 				"type" => "EmailToProspect", 
						// 				"date" => "03-22-2016T22:11:34", 
						// 				"emailAddresses" => [
						// 					"from" => "test12@email.com", 
						// 					"to" => "test1@email.com", 
						// 					"cc" => "test02@email.com", 
						// 					"bcc" => "test02@email.com" 
						// 				], 
						// 				"subject" => "Email Subject", 
						// 				"emailBody" => "Email Body" 
						// 			], 
						// 			[
						// 				"type" => "Show", 
						// 				"date" => "mm/dd/yyyyT03:47:33", 
						// 				"agentId" => "xxxxx", 
						// 				"unitSpaceIds" => "xxxxx,xxxxx", 
						// 				"comments" => "Comments", 
						// 				"eventReasonId" => "xxxxx" 
						// 			], 
						// 			[
						// 				"type" => "Note", 
						// 				"date" => "mm/dd/yyyyT03:47:33", 
						// 				"eventResultId" => "xxxxx", 
						// 				"comments" => "comments" 
						// 			], 
						// 			[
						// 				"type" => "IncomingText", 
						// 				"date" => "04/30/2020T07:20:46", 
						// 				"eventResultId" => "2826839", 
						// 				"comments" => "Incoming_Text" 
						// 			], 
						// 			[
						// 				"type" => "OutgoingText", 
						// 				"date" => "04/30/2020T07:20:46", 
						// 				"eventResultId" => "57428", 
						// 				"comments" => "Outgoing_Text" 
						// 			] 
						// 		] 
						// 	] 
						] 
					] 
				] 
			] 
		] 
	]; 
    
	return $request_data;
	
}

gforms-entrata-addon.php
<?php
/*
	Plugin Name: Gravity Forms Entrata Addon
	Plugin URI: https://elod.in
    Description: Just another plugin to add Entrata feeds to Gravity Forms.
	Version: 1.0.3
    Author: Jon Schroeder
    Author URI: https://elod.in

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.
*/

//* Entrata API documentation: https://www.entrata.com/api/v1/documentation/sendLeads
//* Gforms sample feeds addon: https://github.com/gravityforms/simpleaddon
//* Gforms documentation: https://docs.gravityforms.com/gffeedaddon/

/* Prevent direct access to the plugin */
if ( !defined( 'ABSPATH' ) ) {
    die( "Sorry, you are not allowed to access this page directly." );
}

// Plugin directory
define( 'GFORMS_ENTRATA_ADDON', dirname( __FILE__ ) );

// Define the version of the plugin
define( 'GFORMS_ENTRATA_ADDON_VERSION', '1.0.3' );

add_action( 'gform_loaded', array( 'GF_Entrata_AddOn_Bootstrap', 'load' ), 5 );

class GF_Entrata_AddOn_Bootstrap {

	public static function load() {

		if ( ! method_exists( 'GFForms', 'include_feed_addon_framework' ) ) {
			return;
		}

		require_once( 'lib/class-entrata-addon-feed.php' );

		GFAddOn::register( 'GFEntrataFeedAddon' );
	}

}

function gf_simple_feed_addon() {
	return GFEntrataFeedAddon::get_instance();
}

//* Add the updater
require 'vendor/plugin-update-checker/plugin-update-checker.php';
$myUpdateChecker = Puc_v4_Factory::buildUpdateChecker(
	'https://github.com/jonschr/gforms-entrata-addon',
	__FILE__,
	'gforms-entrata-addon'
);

// Optional: Set the branch that contains the stable release.
$myUpdateChecker->setBranch('master');

gforms-custom-request-maddie.php
<?php
/*
	Plugin Name: Gravity Forms Entrata Addon
	Plugin URI: https://elod.in
    Description: Just another plugin to add Entrata feeds to Gravity Forms.
	Version: 0.1
    Author: Jon Schroeder
    Author URI: https://elod.in

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.
*/

//* Documentation: https://www.entrata.com/api/v1/documentation/sendLeads
//* Request URL is here: https://cardinal.entrata.com/api/v1/leads
//* Test URL is viewable here: https://pipedream.com/sources/dc_Pnu3Q6
//* Test URL endpoint: https://138ed8bde199a4f69a5abd4fc82b86f1.m.pipedream.net

/* Prevent direct access to the plugin */
if ( !defined( 'ABSPATH' ) ) {
    die( "Sorry, you are not allowed to access this page directly." );
}

// Plugin directory
define( 'GFORMS_ENTRATA_ADDON', dirname( __FILE__ ) );

// Define the version of the plugin
define ( 'GFORMS_ENTRATA_ADDON_VERSION', '0.1' );

// Set the propery ID
$property_id = '1058116';

//* Send the request headers for form ID 1
add_filter( 'gform_webhooks_request_headers_1', 'set_headers', 10, 4 );
function set_headers( $request_headers, $feed, $entry, $form ){
	
	//* This is the username:password but base64 encoded
    $request_headers['Authorization'] = 'Basic YnJpbmRsZV9hcGlAY2FyZGluYWw6QnJpbmRsZTUyMjIwMjBe';
    return $request_headers;
}

//* Set up the body for form ID 1
add_filter( 'gform_webhooks_request_data_1', 'modify_data_test_webhook_form', 10, 4 );
function modify_data_test_webhook_form( $request_data, $feed, $entry, $form ){    
	
	$data = $request_data;
	
	// Property details
	$property_id = '1058114';
	$firstname = $data['1.3'];
	$lastname = $data['1.6'];
	$email = $data['2'];
	$phone = $data['3'];
	$message = $data['4'];
	$date = date( 'm/d/Y') . 'T' . date( 'H:i:s' );
	$date = date( 'm/d/Y') . 'T' . '00:00:00';
	$leadsource = '41474'; // 41474 is the Community Website ID
	
	$request_data = [
		"auth" => [
			"type" => "basic" 
		], 
		"requestId" => "15", 
		"method" => [
			"name" => "sendLeads", 
			"params" => [
				"propertyId" => $property_id, 
				"doNotSendConfirmationEmail" => "1", 
				"isWaitList" => "0", 
				"prospects" => [
					"prospect" => [
						[
							"leadSource" => [
								"originatingLeadSourceId" => $leadsource, 
								// "additionalLeadSourceIds" => "xxxx,xxxx,xxxx" 
							], 
							"createdDate" => $date,
							// "createdDate" => "mm/dd/yyyyT03:47:33", 
							// "leasingAgentId" => "318487", 
							"customers" => [
								"customer" => [
									[
										"name" => [
											"firstName" => $firstname,
											"lastName" => $lastname, 
											// "namePrefix" => "Name Prefix", 
											// "middleName" => "Middle Name", 
											// "maidenName" => "Maiden Name", 
											// "nameSuffix" => "Name Suffix" 
										], 
										// "customerRelationshipTypeId" => "XXXX", 
										// "address" => [
										// 	"addressLine1" => "addressLine1", 
										// 	"addressLine2" => "addressLine2", 
										// 	"addressLine3" => "addressLine3", 
										// 	"city" => "city", 
										// 	"state" => "state code", 
										// 	"postalCode" => "postal code" 
										// ], 
										"phone" => [
											"personalPhoneNumber" => $phone, 
											// "cellPhoneNumber" => "xxx-xxx-xxxx", 
											// "officePhoneNumber" => "xxx-xxx-xxxx", 
											// "faxNumber" => "xxx-xxx-xxxx" 
										], 
										"email" => $email,
									] 
								] 
							], 
							"customerPreferences" => [
								// "desiredMoveInDate" => "mm/dd/yyyy", 
								// "desiredFloorplanId" => "xxxx", 
								// "desiredUnitTypeId" => "xxxx", 
								// "desiredUnitId" => "xxxx", 
								// "desiredRent" => [
								// 	"min" => "200", 
								// 	"max" => "300" 
								// ], 
								// "desiredNumBedrooms" => "2", 
								// "desiredNumBathrooms" => "3", 
								// "desiredLeaseTerms" => "18", 
								// "numberOfOccupants" => "1", 
								"comment" => $message, 
							], 
						// 	"events" => [
						// 		"event" => [
						// 			[
						// 				"date" => "mm/dd/yyyyT03:47:33", 
						// 				"callData" => [
						// 					"callFrom" => "xxx-xxx-xxxx", 
						// 					"ringThrough" => "xxx-xxx-xxxx", 
						// 					"callStatus" => "Answer", 
						// 					"duration" => "900", 
						// 					"audioLink" => "URL for the call" 
						// 				] 
						// 			], 
						// 			[
						// 				"type" => "Appointment", 
						// 				"date" => "mm/dd/yyyyT22:11:34", 
						// 				"appointmentDate" => "mm/dd/yyyy", 
						// 				"timeFrom" => "11:00am", 
						// 				"timeTo" => "11:30am", 
						// 				"eventReasons" => "Appointment Title", 
						// 				"comments" => "Appointment Description" 
						// 			], 
						// 			[
						// 				"type" => "EmailFromProspect", 
						// 				"date" => "mm/dd/yyyyT22:11:34", 
						// 				"emailAddresses" => [
						// 					"from" => "test1@email.com", 
						// 					"to" => "test11@email.com", 
						// 					"cc" => "test11@email.com", 
						// 					"bcc" => "test11@email.com" 
						// 				], 
						// 				"subject" => "Email Subject", 
						// 				"emailBody" => "Email Body" 
						// 			], 
						// 			[
						// 				"type" => "EmailToProspect", 
						// 				"date" => "03-22-2016T22:11:34", 
						// 				"emailAddresses" => [
						// 					"from" => "test12@email.com", 
						// 					"to" => "test1@email.com", 
						// 					"cc" => "test02@email.com", 
						// 					"bcc" => "test02@email.com" 
						// 				], 
						// 				"subject" => "Email Subject", 
						// 				"emailBody" => "Email Body" 
						// 			], 
						// 			[
						// 				"type" => "Show", 
						// 				"date" => "mm/dd/yyyyT03:47:33", 
						// 				"agentId" => "xxxxx", 
						// 				"unitSpaceIds" => "xxxxx,xxxxx", 
						// 				"comments" => "Comments", 
						// 				"eventReasonId" => "xxxxx" 
						// 			], 
						// 			[
						// 				"type" => "Note", 
						// 				"date" => "mm/dd/yyyyT03:47:33", 
						// 				"eventResultId" => "xxxxx", 
						// 				"comments" => "comments" 
						// 			], 
						// 			[
						// 				"type" => "IncomingText", 
						// 				"date" => "04/30/2020T07:20:46", 
						// 				"eventResultId" => "2826839", 
						// 				"comments" => "Incoming_Text" 
						// 			], 
						// 			[
						// 				"type" => "OutgoingText", 
						// 				"date" => "04/30/2020T07:20:46", 
						// 				"eventResultId" => "57428", 
						// 				"comments" => "Outgoing_Text" 
						// 			] 
						// 		] 
						// 	] 
						] 
					] 
				] 
			] 
		] 
	]; 
    
	return $request_data;
	
}


lib/class-entrata-addon-feed.php
<?php

//* Based on the simple feed addon for Gravity forms sample code
//* https://github.com/gravityforms/simplefeedaddon/blob/master/class-gfsimplefeedaddon.php

GFForms::include_feed_addon_framework();

class GFEntrataFeedAddon extends GFFeedAddOn {

	protected $_version = GFORMS_ENTRATA_ADDON_VERSION;
	protected $_min_gravityforms_version = '1.9.16';
	protected $_slug = 'gforms-entrata-addon';
	protected $_path = 'gforms-entrata-addon/gforms-entrata-addon.php';
	protected $_full_path = __FILE__;
	protected $_title = 'Gravity Forms Entrata Feed';
	protected $_short_title = 'Entrata';

	private static $_instance = null;

	/**
	 * Get an instance of this class.
	 *
	 * @return GFEntrataFeedAddon
	 */
	public static function get_instance() {
		if ( self::$_instance == null ) {
			self::$_instance = new GFEntrataFeedAddon();
		}

		return self::$_instance;
	}

	/**
	 * Plugin starting point. Handles hooks, loading of language files and PayPal delayed payment support.
	 */
	public function init() {

		parent::init();

		$this->add_delayed_payment_support(
			array(
				'option_label' => esc_html__( 'Subscribe contact to service x only when payment is received.', 'entrata_gforms_addon' )
			)
		);

	}


	// # FEED PROCESSING -----------------------------------------------------------------------------------------------

	/**
	 * Process the feed e.g. subscribe the user to a list.
	 *
	 * @param array $feed The feed object to be processed.
	 * @param array $entry The entry object currently being processed.
	 * @param array $form The form object currently being processed.
	 *
	 * @return bool|void
	 */
	public function process_feed( $feed, $entry, $form ) {

        // Retrieve the name => value pairs for all fields mapped in the 'mappedFields' field map.
        $field_map = $this->get_field_map_fields( $feed, 'mappedFields' );

		// Loop through the fields from the field map setting building an array of values to be passed to the third-party service.
		$merge_vars = array();
		foreach ( $field_map as $name => $field_id ) {

			// Get the field value for the specified field id
			$merge_vars[ $name ] = $this->get_field_value( $form, $entry, $field_id );

        }
                
        $requesturl = $feed['meta']['entrata_requesturl'];
        $username = $feed['meta']['entrata_username'];
        $password = $feed['meta']['entrata_password'];
        $property_id = $feed['meta']['entrata_propertyid'];
        $leadsource = $feed['meta']['entrata_leadsource'];
        $firstname = $merge_vars['first_name'];
        $lastname = $merge_vars['last_name'];
        $email = $merge_vars['email'];
        $phone = $merge_vars['phone'];
        $message = $merge_vars['message'];
		// $date = date('m/d/Y', mktime(0, 0, 0, date("m"), date("d") -1, date("Y"))) . 'T' . '00:00:00';
		// "createdDate": "mm/dd/yyyyT03:47:33",
		
		// $date = date('m/d/Y', mktime(0, 0, 0, date("m"), date("d") -1, date("Y"))) . 'T' . '00:00:00';
		
		date_default_timezone_set('America/Denver');
		$date = sprintf( '%sT%s', date( 'm/d/Y' ), date( 'G:i:s' ) );
        
        $body = [
            "auth" => [
                "type" => "basic" 
            ], 
            "requestId" => "15", 
            "method" => [
                "name" => "sendLeads", 
                "params" => [
                    "propertyId" => $property_id, 
                    "doNotSendConfirmationEmail" => "1", 
                    "isWaitList" => "0", 
                    "prospects" => [
                        "prospect" => [
                            [
                                "leadSource" => [
                                    "originatingLeadSourceId" => $leadsource, 
                                ], 
                                "createdDate" => $date,
                                "customers" => [
                                    "customer" => [
                                        [
                                            "name" => [
                                                "firstName" => $firstname,
                                                "lastName" => $lastname, 
                                            ], 
                                            "phone" => [
                                                "personalPhoneNumber" => $phone, 
                                            ], 
                                            "email" => $email,
                                        ] 
                                    ] 
                                ], 
                                "customerPreferences" => [
                                    "comment" => $message, 
                                ], 
                            ] 
                        ] 
                    ] 
                ] 
            ] 
        ]; 
        
        $body = json_encode( $body );
        
        //* Set up the auth part of the request
        $auth = $username . ':' . $password;
        $auth = base64_encode( $auth );
        
        $request = array(
            'body'        => $body,
            'headers'     => array(
                'Content-Type'      => 'application/json',
                'Authorization'     => 'Basic ' . $auth,
            ),                
            'timeout'     => 60,
            'redirection' => 5,
            'blocking'    => true,
            'httpversion' => '1.0',
            'sslverify'   => false,
            'data_format' => 'body',
        );
        
        $this->log_debug( 'Outgoing request body: ' . $body );
        
        $response = wp_remote_post( $requesturl, $request );
        
        // Log error or success based on response.
		if ( is_wp_error( $response ) ) {
			$this->add_feed_error( sprintf( esc_html__( 'Webhook was not successfully executed. %s (%d)', 'gforms_entrata' ), $response->get_error_message(), $response->get_error_code() ), $feed, $entry, $form );
		} else {
			$this->log_debug( sprintf( '%s(): Webhook successfully executed. code: %s; body: %s', __METHOD__, wp_remote_retrieve_response_code( $response ), wp_remote_retrieve_body( $response ) ) );
		}
        
	}

	/**
	 * Custom format the phone type field values before they are returned by $this->get_field_value().
	 *
	 * @param array $entry The Entry currently being processed.
	 * @param string $field_id The ID of the Field currently being processed.
	 * @param GF_Field_Phone $field The Field currently being processed.
	 *
	 * @return string
	 */
	public function get_phone_field_value( $entry, $field_id, $field ) {

		// Get the field value from the Entry Object.
		$field_value = rgar( $entry, $field_id );

		// If there is a value and the field phoneFormat setting is set to standard reformat the value.
		if ( ! empty( $field_value ) && $field->phoneFormat == 'standard' && preg_match( '/^\D?(\d{3})\D?\D?(\d{3})\D?(\d{4})$/', $field_value, $matches ) ) {
			$field_value = sprintf( '%s-%s-%s', $matches[1], $matches[2], $matches[3] );
		}

		return $field_value;
	}

	/**
	 * Configures the settings which should be rendered on the feed edit page in the Form Settings > Simple Feed Add-On area.
	 *
	 * @return array
	 */
	public function feed_settings_fields() {
		return array(
			array(
				'title'  => esc_html__( 'Entrata sendLeads API details', 'entrata_gforms_addon' ),
				'fields' => array(
                    array(
						'label'   => esc_html__( 'Feed name', 'entrata_gforms_addon' ),
						'type'    => 'text',
						'name'    => 'feedName',
						'class'   => 'small',
                    ),
                    array(
						'label'   => esc_html__( 'Request URL', 'entrata_gforms_addon' ),
						'type'    => 'text',
                        'name'    => 'entrata_requesturl',
                        'tooltip'   => esc_html__( 'E.g. https://YOURDOMAIN.entrata.com/api/v1/leads', 'gforms_entrata' ),
						'class'   => 'medium',
					),
                    array(
						'type'      => 'text',
						'name'      => 'entrata_username',
						'label'     => esc_html__( 'Entrata username', 'gforms_entrata' ),
						'class'		=>  'medium',
					),
					array(
						'type'      => 'text',
						'name'      => 'entrata_password',
						'label'     => esc_html__( 'Entrata password', 'gforms_entrata' ),
						'class'		=>  'medium',
					),
					array(
						'type'      => 'text',
						'name'      => 'entrata_propertyid',
						'label'     => esc_html__( 'Entrata property ID', 'gforms_entrata' ),
						'tooltip'   => esc_html__( 'This should be a 7-digit number, e.g. 1234567', 'gforms_entrata' ),
						'class'		=>  'medium',
                    ),
                    array(
						'type'      => 'text',
						'name'      => 'entrata_leadsource',
						'label'     => esc_html__( 'Lead source ID', 'gforms_entrata' ),
						'tooltip'   => esc_html__( 'This should be a 5-digit number, e.g. 12345', 'gforms_entrata' ),
						'class'		=>  'medium',
					),
					array(
						'name'      => 'mappedFields',
						'label'     => esc_html__( 'Map Fields', 'entrata_gforms_addon' ),
						'type'      => 'field_map',
						'field_map' => array(
							array(
								'name'     => 'first_name',
								'label'    => esc_html__( 'First name', 'entrata_gforms_addon' ),
								'required' => 0,
                            ),
                            array(
								'name'     => 'last_name',
								'label'    => esc_html__( 'Last name', 'entrata_gforms_addon' ),
								'required' => 0,
                            ),
                            array(
								'name'       => 'email',
								'label'      => esc_html__( 'Email', 'entrata_gforms_addon' ),
								'required'   => 0,
								'field_type' => 'email',
							),
							array(
								'name'       => 'phone',
								'label'      => esc_html__( 'Phone', 'entrata_gforms_addon' ),
								'required'   => 0,
								'field_type' => 'phone',
                            ),
                            array(
								'name'       => 'message',
								'label'      => esc_html__( 'Message', 'entrata_gforms_addon' ),
								'required'   => 0,
							),
						),
					),
					// array(
					// 	'name'           => 'condition',
					// 	'label'          => esc_html__( 'Condition', 'entrata_gforms_addon' ),
					// 	'type'           => 'feed_condition',
					// 	'checkbox_label' => esc_html__( 'Enable Condition', 'entrata_gforms_addon' ),
					// 	'instructions'   => esc_html__( 'Process this simple feed if', 'entrata_gforms_addon' ),
					// ),
				),
			),
		);
	}

	/**
	 * Configures which columns should be displayed on the feed list page.
	 *
	 * @return array
	 */
	public function feed_list_columns() {
		return array(
			'feedName'  => esc_html__( 'Feed Name', 'entrata_gforms_addon' ),
            'entrata_propertyid' => esc_html__( 'Property ID', 'entrata_gforms_addon' ),
            'entrata_username' => esc_html__( 'Entrata User', 'entrata_gforms_addon' ),
            'entrata_password' => esc_html__( 'Entrata Pass', 'entrata_gforms_addon' ),
		);
	}
}